Lib/dataclasses.py

```diff
@@ -0,0 +1,776 @@
+    # Return the __init__ parameter string for this field.
+    #  For example, the equivalent of 'x:int=3' (except instead of 'int',
+    #  reference a variable set to int, and instead of '3', reference a
+    #  variable set to 3).
+    if f.default is _MISSING and f.default_factory is _MISSING:
+        # There's no default, and no default_factory, just
+        #  output the variable name and type.
+        default = ''
+    elif f.default is not _MISSING:
+        # There's a default, this will be the name that's used to look it up.
+        default = f'=_dflt_{f.name}'
+    elif f.default_factory is not _MISSING:
+        # There's a factory function. Set a marker.
+        default = '=_HAS_DEFAULT_FACTORY'
+    return f'{f.name}:_type_{f.name}{default}'
+
+
+def _init_fn(fields, frozen, has_post_init, self_name):
+    # fields contains both real fields and InitVar pseudo-fields.
+
+    # Make sure we don't have fields without defaults following fields
+    #  with defaults.  This actually would be caught when exec-ing the
+    #  function source code, but catching it here gives a better error
+    #  message, and future-proofs us in case we build up the function
+    #  using ast.
+    seen_default = False
+    for f in fields:
+        # Only consider fields in the __init__ call.
+        if f.init:
+            if not (f.default is _MISSING and f.default_factory is _MISSING):
+                seen_default = True
+            elif seen_default:
+                raise TypeError(f'non-default argument {f.name!r} '
+                                'follows default argument')
+
+    globals = {'_MISSING': _MISSING,
+               '_HAS_DEFAULT_FACTORY': _HAS_DEFAULT_FACTORY}
+
+    body_lines = []
+    for f in fields:
+        # Do not initialize the pseudo-fields, only the real ones.
+        line = _field_init(f, frozen, globals, self_name)
+        if line is not None:
+            # line is None means that this field doesn't require
+            #  initialization. Just skip it.
+            body_lines.append(line)
+
+    # Does this class have a post-init function?
+    if has_post_init:
+        params_str = ','.join(f.name for f in fields
+                              if f._field_type is _FIELD_INITVAR)
+        body_lines += [f'{self_name}.{_POST_INIT_NAME}({params_str})']
+
+    # If no body lines, use 'pass'.
+    if len(body_lines) == 0:
+        body_lines = ['pass']
+
+    locals = {f'_type_{f.name}': f.type for f in fields}
+    return _create_fn('__init__',
+                      [self_name] +[_init_param(f) for f in fields if f.init],
+                      body_lines,
+                      locals=locals,
+                      globals=globals,
+                      return_type=None)
+
+
+def _repr_fn(fields):
+    return _create_fn('__repr__',
+                      ['self'],
+                      ['return self.__class__.__qualname__ + f"(' +
+                       ', '.join([f"{f.name}={{self.{f.name}!r}}"
```

Comment:
Could you refactor this? Extract the inner expression as a variable.

Reply:
It seems clear to me, but I'll look at it.